contrib/jms/src/main/java/com/google/cloud/dataflow/contrib/jms/JmsIO.java

```diff
@@ -0,0 +1,532 @@
+            }
+
+            @Override
+            protected Coder<T> getDefaultOutputCoder() {
+                return coder;
+            }
+
+            public ConnectionFactory getConnectionFactory() {
+                return connectionFactory;
+            }
+
+            public String getQueue() {
+                return queue;
+            }
+
+            public String getTopic() {
+                return topic;
+            }
+
+            public Coder<T> getCoder() {
+                return coder;
+            }
+
+            public int getMaxNumMessages() {
+                return maxNumMessages;
+            }
+
+            private class JmsReader extends DoFn<Void, T> {
+
+                @Override
+                public void processElement(ProcessContext c) throws Exception {
+                    LOG.debug("Connecting to the JMS broker");
+                    Connection connection = connectionFactory.createConnection();
+                    connection.start();
+                    Session session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+                    MessageConsumer consumer;
+                    if (queue != null) {
+                        consumer = session.createConsumer(session.createQueue(queue));
+                    } else {
+                        consumer = session.createConsumer(session.createTopic(topic));
+                    }
+
+                    List<TextMessage> messages = new ArrayList<>();
+
+                    while ((getMaxNumMessages() == 0 ||
+                            messages.size() <= getMaxNumMessages())) {
+                        LOG.debug("Consuming JMS message");
+                        TextMessage message = (TextMessage) consumer.receive();
+                        LOG.debug("Adding JMS message to the list ({})", messages.size());
+                        messages.add(message);
+                    }
+
+                    consumer.close();
+                    session.close();
+                    connection.stop();
+                    connection.close();
+
+                    for (TextMessage message : messages) {
+                        c.output(CoderUtils.decodeFromByteArray(coder,
+                                message.getText().getBytes()));
+                    }
+                }
+
+            }
+
+
+        }
+
+        /** Disallow construction of utility class. */
+        private Read() {}
+
+    }
+
+    /** Disallow construction of utility class. */
+    private JmsIO() {}
+
+    /**
+     * A {@link PTransform} that continuously writes a {@link PCollection} of {@link String Strings}
+     * to a JMS broker on the specified destination.
+     */
+    public static class Write {
+
+        /**
+         * Creates a transform that produces to a JMS broker with the given step name.
+         *
+         * @param name the step name
+         * @return the Bound
+         */
+        public static Bound<String> named(String name) {
```

Comment:
similar recommendations w.r.t dropping `Write.Bound` and just using `Write` instead, dropping `named`, adding `withX` as appropriate.